Documentation/ExtensionArchitecture/HowTo/Localization/Php.rst

```diff
@@ -0,0 +1,92 @@
+.. include:: /Includes.rst.txt
+.. index::
+   Localization; PHP
+.. _extension-localization-php:
+
+====================
+Localization in PHP
+====================
+
+Sometimes you have to localize a string in PHP code, for
+example inside of a controller or a ViewHelper.
+
+You can get the translation of a string into the current language via:
+
+..  code-block:: php
+    :caption: EXT:my_extension/Classes/SomeClass.php
+
+    $translatedString = $GLOBALS['LANG']->sL($translationKey);
```

Comment:
$GLOBALS['LANG'] is mostly backend and cannot be used in frontend: When used in frontend and a backend user is logged in, the backend user language is used for lacalization.
See also: https://github.com/TYPO3/typo3/blob/main/typo3/sysext/core/Classes/Localization/LanguageService.php#L28

Mostly I use LocalizationUtility from Extbase. Don't know if this is best choice.

Reply:
However $GLOBALS['LANG'] is used by the translation ViewHelper in the Core, which in turn is used in the frontend, so I figured that this is best practice?

Reply:
Strange, I really had the problem an hour ago, that GLOBALS['LANG'] used the backend user language (custom "gsw_CH") for frontend ("de" and "en"). After switching to LocalizationUtility it worked correctly.

Reply:
Maybe it is only available in the View and therefore for ViewHelpers. The class comments states "Some Cirumstances"

Reply:
And I am getting an error when not logged-in into backend: Call to a member function sL() on null. I just use a plain user funtion for testing. Maybe translate viewhelper initialises something before using?

Reply:
I remember vagely, that when the Admin Panel is activated, $GLOBALS['LANG'] works out of the box in frontend. That is obviously  then not the case without logged-in user.

Reply:
Here we also have the suggestion to use the Language Service: **https://docs.typo3.org/m/typo3/reference-coreapi/main/en-us/ApiOverview/Internationalization/XliffApi.html#label-access-in-php** However it does not state where we are at that moment :(

Reply:
"In the backend" -> this is correct. But not in frontend context.

Reply:
TSFE has a `sL` method which calls then ... tadaaa ... LanguageService:
https://github.com/TYPO3/typo3/blob/main/typo3/sysext/frontend/Classes/Controller/TypoScriptFrontendController.php#L2513

And the language service is initialized in TSFE the LanguageServiceFactory:
https://github.com/TYPO3/typo3/blob/main/typo3/sysext/frontend/Classes/Controller/TypoScriptFrontendController.php#L2524

This looks pretty good:
https://github.com/TYPO3/typo3/blob/main/typo3/sysext/core/Classes/Localization/LanguageServiceFactory.php#L46

But don't know if it is best practise to use that factory.

Reply:
Maybe just use the TSFE. But TSFE is not injectable anymore since v12 (it has a state). In user functions you can use the available content object renderer, in controllers it should be also available without global variable.

Reply:
Would you like to take over? For me it is starting to be to complicate

Reply:
I can make a draft for using TSFE in different contexts, but would like someone from core team to look over that before merging. Don't know if I can manage this in the next days. I'll try.

Then we can reference that chapter and suggest `TSFE->sL` for non-Extbase context and `LocalizationUtility` for Extbase context.